db/migrate/20211201221028_migrate_email_to_normalized_email.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+class MigrateEmailToNormalizedEmail < ActiveRecord::Migration[6.1]
+
+  # minimize locking on user_email table
+  disable_ddl_transaction!
+
+  def up
+
+    min, max, total = DB.query_single "SELECT MIN(id), MAX(id), COUNT(*) FROM user_emails"
+    # scaling is needed to compensate for "holes" where records were deleted
+    # and pathological cases where for some reason id 100_000_000 and 0 exist
+    scaling = total.to_f / (max - min).to_f
+
+    batch_size = 100_000
+
+    enum = Enumerator.new do |yielder|
```

Comment:
I see how this is working, is this just to avoid the slow `LIMIT / OFFSET` query?

Reply:
It would be cool if we could somehow make this more generic if we need to use it later on, e.g. in mini_sql. `DB.query_batch("SELECT MIN(id), MAX(id), COUNT(*) FROM user_emails").each do |low_id, high_id|` or something?

Reply:
yeah some generic pattern here that leans on TABLESAMPLE or something would be very cool. you need to sample in an ordered way as well, then adjust for different sized gaps or just work through all ids first and find the "correct" bounds without selecting millions of ids and shipping to the app.

Reply:
yeah limit / offset is going to be slower.